desc/compute/_metric.py

```diff
@@ -686,6 +686,888 @@ def _g_sup_tz(params, transforms, profiles, data, **kwargs):
     return data
 
 
+@register_compute_fun(
+    name="g^rr_r",
+    label="g^{\\rho}{\\rho}_{\\rho}",
+    units="m^-2",
+    units_long="inverse square meters",
+    description="Radial/Radial element of contravariant metric tensor, "
+    + "first radial derivative",
+    dim=1,
+    params=[],
+    transforms={},
+    profiles=[],
+    coordinates="rtz",
+    data=[
+        "e^rho",
+        "e^rho",
+        "sqrt(g)_r",
+        "sqrt(g)",
+        "e_rho",
+        "e_rho",
+        "e_theta",
+        "e_rho_r",
+        "e_rho_r",
+        "e_theta_r",
+    ],
+)
+def _g_sup_rr_r(params, transforms, profiles, data, **kwargs):
```

Comment:
this seems incorrect to me.
let $r=\rho, t=\theta, z=\zeta$
then
$e^r = e_t \times e_z / \sqrt{g}$
$g^{rr} = e^r \cdot e^r$
$\partial_r (g^{rr}) = 2 \partial_r (e^r) \cdot e^r$
$\partial_r (e^r) = \frac{ \partial_re_t \times e_z + e_t \times \partial_re_z }{\sqrt{g}} - \frac{e_t \times e_z}{(\sqrt{g})^2}$ $\partial_r (\sqrt{g})$

what you have looks like it is $\partial_r g^{zz}$

Reply:
or maybe I am just getting my derivatives wrong, but I definitely think that the covariant vectors here should be $e_t$ and $e_z$, not $e_r$ and $e_t$

Reply:
This is the derivation: 
![image](https://user-images.githubusercontent.com/22567458/228951758-7f9476cc-52ed-4a7f-a0fe-9d78ab20e997.png)

The code was basically generated by substituting for i,j,k,l etc.


```
lab = ["rho", "theta", "zeta"]
tex = ["""{\\\\""" + foo + "}" for foo in lab]
cha = ["r", "t", "z"]
des = ["radial", "poloidal", "toroidal"]
Des = ["Radial", "Poloidal", "Toroidal"]



cnt = 0

for l in range(3):
    for i in range(3):
        for j in range(i,3):
            cnt += 1
            k = list({0,1,2}.difference({i,j}))[0]
            s = (
f"""
@register_compute_fun(
    name="g^{cha[i]}{cha[j]}_{cha[l]}",
    label="g^{tex[i]}{tex[j]}_{tex[l]}",
    units="m^{-2}",
    units_long="inverse square meters",
    description="{Des[i]}/{Des[j]} element of contravariant metric tensor, "
    + "first {des[l]} derivative",
    dim=1,
    params=[],
    transforms={{}},
    profiles=[],
    coordinates="rtz",
    data=["e^{lab[i]}", "e^{lab[j]}", "sqrt(g)_{cha[l]}", "sqrt(g)", "e_{lab[i]}", "e_{lab[j]}", "e_{lab[k]}", "e_{lab[i]}_{cha[l]}", "e_{lab[j]}_{cha[l]}", "e_{lab[k]}_{cha[l]}"],
)
def _g_sup_{cha[i]}{cha[j]}_{cha[l]}(params, transforms, profiles, data, **kwargs):
    data["g^{cha[i]}{cha[j]}_{cha[l]}"] = data["sqrt(g)_{cha[l]}"]/data["sqrt(g)"]**2*dot(data["e^{lab[j]}"], cross(data["e_{lab[j]}"], data["e_{lab[k]}"])) + 1/data["sqrt(g)"]*dot(data["e^{lab[j]}"], cross(data["e_{lab[j]}_{cha[l]}"], data["e_{lab[k]}"])) + 1/data["sqrt(g)"]*dot(data["e^{lab[j]}"], cross(data["e_{lab[j]}"], data["e_{lab[k]}_{cha[l]}"])) + data["sqrt(g)_{cha[l]}"]/data["sqrt(g)"]**2*dot(data["e^{lab[i]}"], cross(data["e_{lab[k]}"], data["e_{lab[i]}"])) + 1/data["sqrt(g)"]*dot(data["e^{lab[i]}"], cross(data["e_{lab[k]}_{cha[l]}"], data["e_{lab[i]}"])) + 1/data["sqrt(g)"]*dot(data["e^{lab[i]}"], cross(data["e_{lab[k]}"], data["e_{lab[i]}_{cha[l]}"]))
    return data
"""
            )
            print(s)

Reply:
It might be my code generation is wrong when i==j but can't figure it out right now. I'll take another look tomorrow, if something jumps out at you let me know

Reply:
This should be correct, I just changed the labels so that the covariant vectors are defined as the cyclic pair starting from whatever the contravariant metric's indices are 

```python
lab = ["rho", "theta", "zeta"]
tex = ["""{\\\\""" + foo + "}" for foo in lab]
cha = ["r", "t", "z"]
des = ["radial", "poloidal", "toroidal"]
Des = ["Radial", "Poloidal", "Toroidal"]



cnt = 0

for l in range(3):
    for i in range(3):
        for j in range(i,3):
            cnt += 1
            k = list({0,1,2}.difference({i,j}))[0]
            s = (
f"""
@register_compute_fun(
    name="g^{cha[i]}{cha[j]}_{cha[l]}",
    label="g^{tex[i]}{tex[j]}_{tex[l]}",
    units="m^{-2}",
    units_long="inverse square meters",
    description="{Des[i]}/{Des[j]} element of contravariant metric tensor, "
    + "first {des[l]} derivative",
    dim=1,
    params=[],
    transforms={{}},
    profiles=[],
    coordinates="rtz",
    data=["e^{lab[i]}", "e^{lab[j]}", "sqrt(g)_{cha[l]}", "sqrt(g)", "e_{lab[i]}", "e_{lab[j]}", "e_{lab[k]}", "e_{lab[i]}_{cha[l]}", "e_{lab[j]}_{cha[l]}", "e_{lab[k]}_{cha[l]}"],
)
def _g_sup_{cha[i]}{cha[j]}_{cha[l]}(params, transforms, profiles, data, **kwargs):
    data["g^{cha[i]}{cha[j]}_{cha[l]}"] = data["sqrt(g)_{cha[l]}"]/data["sqrt(g)"]**2*dot(data["e^{lab[j]}"], cross(data["e_{lab[(i+1)%3]}"], data["e_{lab[(i+2)%3]}"])) + 1/data["sqrt(g)"]*dot(data["e^{lab[j]}"], cross(data["e_{lab[(i+1)%3]}_{cha[l]}"], data["e_{lab[(i+2)%3]}"])) + 1/data["sqrt(g)"]*dot(data["e^{lab[j]}"], cross(data["e_{lab[(j+1)%3]}"], data["e_{lab[(i+2)%3]}_{cha[l]}"])) + data["sqrt(g)_{cha[l]}"]/data["sqrt(g)"]**2*dot(data["e^{lab[i]}"], cross(data["e_{lab[(j+1)%3]}"], data["e_{lab[(j+2)%3]}"])) + 1/data["sqrt(g)"]*dot(data["e^{lab[i]}"], cross(data["e_{lab[(j+1)%3]}_{cha[l]}"], data["e_{lab[(j+2)%3]}"])) + 1/data["sqrt(g)"]*dot(data["e^{lab[i]}"], cross(data["e_{lab[(j+1)%3]}"], data["e_{lab[(j+2)%3]}_{cha[l]}"]))
    return data
"""
            )
            print(s)
```

Reply:
OK yeah I see the issue. I also had to modify the data dependency bit. For reference here's the final code generator:

```python
import re
lab = ["rho", "theta", "zeta"]
tex = ["""{\\\\""" + foo + "}" for foo in lab]
cha = ["r", "t", "z"]
des = ["radial", "poloidal", "toroidal"]
Des = ["Radial", "Poloidal", "Toroidal"]



cnt = 0

for l in range(3):
    for i in range(3):
        for j in range(i,3):
            cnt += 1
            k = list({0,1,2}.difference({i,j}))[0]
            s = (
f"""
@register_compute_fun(
    name="g^{cha[i]}{cha[j]}_{cha[l]}",
    label="g^{tex[i]}{tex[j]}_{tex[l]}",
    units="m^{-2}",
    units_long="inverse square meters",
    description="{Des[i]}/{Des[j]} element of contravariant metric tensor, "
    + "first {des[l]} derivative",
    dim=1,
    params=[],
    transforms={{}},
    profiles=[],
    coordinates="rtz",
    """
)
            src=(
f"""
def _g_sup_{cha[i]}{cha[j]}_{cha[l]}(params, transforms, profiles, data, **kwargs):
    data["g^{cha[i]}{cha[j]}_{cha[l]}"] = data["sqrt(g)_{cha[l]}"]/data["sqrt(g)"]**2*dot(data["e^{lab[j]}"], cross(data["e_{lab[(i+1)%3]}"], data["e_{lab[(i+2)%3]}"])) + 1/data["sqrt(g)"]*dot(data["e^{lab[j]}"], cross(data["e_{lab[(i+1)%3]}_{cha[l]}"], data["e_{lab[(i+2)%3]}"])) + 1/data["sqrt(g)"]*dot(data["e^{lab[j]}"], cross(data["e_{lab[(j+1)%3]}"], data["e_{lab[(i+2)%3]}_{cha[l]}"])) + data["sqrt(g)_{cha[l]}"]/data["sqrt(g)"]**2*dot(data["e^{lab[i]}"], cross(data["e_{lab[(j+1)%3]}"], data["e_{lab[(j+2)%3]}"])) + 1/data["sqrt(g)"]*dot(data["e^{lab[i]}"], cross(data["e_{lab[(j+1)%3]}_{cha[l]}"], data["e_{lab[(j+2)%3]}"])) + 1/data["sqrt(g)"]*dot(data["e^{lab[i]}"], cross(data["e_{lab[(j+1)%3]}"], data["e_{lab[(j+2)%3]}_{cha[l]}"]))
    return data
"""
)
            pattern = r"data\[(.*?)\]"
            variables = re.findall(pattern, src)
            variables = [s.replace("'", "").replace('"', "") for s in variables]
            variables = sorted(list(set(variables[1:])))
            s += f"data={variables}\n)"
            print(s+src)